token-repository/src/main/scala/io/renku/tokenrepository/repository/init/TokensMigrator.scala

```diff
@@ -0,0 +1,159 @@
+    tokenValidator:            TokenValidator[F],
+    tokenRemover:              TokenRemover[F],
+    projectAccessTokenCreator: ProjectAccessTokenCreator[F],
+    associationPersister:      AssociationPersister[F],
+    queriesExecTimes:          LabeledHistogram[F],
+    retryInterval:             Duration = 5 seconds
+) extends DbClient[F](Some(queriesExecTimes))
+    with DBMigration[F]
+    with TokenRepositoryTypeSerializers {
+
+  private val logPrefix = "token migration:"
+
+  import associationPersister._
+  import fs2.Stream
+  import io.renku.db.SqlStatement
+  import projectAccessTokenCreator._
+  import skunk.Void
+  import skunk.implicits._
+  import tokenCrypto._
+  import tokenRemover._
+  import tokenValidator._
+
+  override def run(): F[Unit] =
+    Stream
+      .emit(())
+      .repeat
+      .evalMap(_ => findTokenWithoutDates)
+      .takeThrough(_.nonEmpty)
+      .flatMap(maybeProjectAndToken => Stream.emits(maybeProjectAndToken.toList))
+      .evalMap { case (proj, encToken) => decryptOrDelete(proj, encToken) }
+      .flatMap(maybeProjectAndToken => Stream.emits(maybeProjectAndToken.toList))
+      .evalMap { case (proj, token) => deleteWhenInvalidWithRetry(proj, token) }
+      .flatMap(maybeProjectAndToken => Stream.emits(maybeProjectAndToken.toList))
+      .evalMap { case (proj, token) => createTokenWithRetry(proj, token) }
+      .flatMap(maybeProjectAndTokenInfo => Stream.emits(maybeProjectAndTokenInfo.toList))
+      .evalMap { case (proj, newTokenInfo) => encrypt(newTokenInfo.token).map(enc => (proj, newTokenInfo, enc)) }
+      .evalTap { case (proj, newTokenInfo, encToken) => persistWithRetry(proj, newTokenInfo, encToken) }
+      .evalMap { case (proj, _, _) => Logger[F].info(show"$logPrefix $proj token created") }
+      .compile
+      .drain
+
+  private def findTokenWithoutDates: F[Option[(Project, EncryptedAccessToken)]] = SessionResource[F].useK {
+    measureExecutionTime {
+      findTokenQuery.build(_.option)
+    }
+  }
+
+  private def findTokenQuery: SqlStatement.Select[F, Void, (Project, EncryptedAccessToken)] =
+    SqlStatement
+      .named[F]("find non-migrated token")
+      .select[Void, Project ~ EncryptedAccessToken](
+        sql"""
+        SELECT project_id, project_path, token
+        FROM projects_tokens
+        WHERE expiry_date IS NULL"""
+          .query(projectIdDecoder ~ projectPathDecoder ~ encryptedAccessTokenDecoder)
+          .map { case (id: projects.Id) ~ (path: projects.Path) ~ (token: EncryptedAccessToken) =>
+            Project(id, path) -> token
+          }
+      )
+      .arguments(Void)
+
+  private def decryptOrDelete(project: Project, encToken: EncryptedAccessToken): F[Option[(Project, AccessToken)]] =
+    decrypt(encToken)
+      .map(t => Option(project -> t))
+      .recoverWith { case ex: Throwable =>
+        Logger[F].error(ex)(show"$logPrefix $project token decryption failed; deleting") >>
+          delete(project.id) >>
```

Comment:
Is the project deleted here or the token?

Reply:
The token. I'll add some context here for better readability.